Create dashboard and DataWrapper

emilom opened the pull request:
Closes #30
Closes #51

Commit:
Add index.css to index.tsx

Commit:
Add default css

Commit:
Remove .keep

Commit:
Make dashboard

Commit:
Make button

Commit:
Add story for dashboard

Commit:
Add dashboard types

Commit:
Add mockdata for dashboard list

Commit:
Add apollo client

Commit:
Add apollo cache

Commit:
Make dashboard more dynamic

Commit:
Add metAPI query

Commit:
Merge branch 'master' of https://github.com/IT2901-SINTEF01/frontend into 30-create-dashboard

Commit:
update with new component

Commit:
Make apollo client use proxy

Commit:
Change types to match query results

Commit:
Change type name

Commit:
Make data wrapper

Commit:
Change storie to match types

Commit:
Move formatter to utils

Commit:
Comment out function for potential later use

blauks commented on src/index.css:
```diff
@@ -0,0 +1,8 @@
+
+body {
+    margin: 0;
+}
```
```suggestion
}

```

FredrikAugust replied:
```suggestion
}

```

As preferred by Dr. @blauks

emilom commented:
This is mostly done,  storybook is coming. Had some problems with apollo provider in storybook. @blauks and I will finish this today

blauks committed:
Update src/index.css

FredrikAugust commented on src/utils/dashboardList.ts:
```diff
@@ -0,0 +1,11 @@
+import { DashboardItemSize, DashboardItemInfo } from '../types/dashboard';
+
+/*
+export const generateRandomList = (size: number): DashboardItemInfo[] =>
+    new Array(size).fill(undefined).map(() => ({
+        size: [DashboardItemSize.SMALL, DashboardItemSize.MEDIUM, DashboardItemSize.LARGE][
+            Math.floor(Math.random() * 3)
+        ],
+        withText: false,
+    }));
+*/
```
Try to avoid commented-out code in PRs.

FredrikAugust replied:
Also check commit comments for more information

FredrikAugust commented on src/types/dashboard.ts:
```diff
@@ -0,0 +1,15 @@
+//Can be expanded with visualisation method if needed.
+//Hva vi må legge til. Text. Title. Visualisation method. Query. //Source.
```
:gb:

blauks replied:
The norwegian comment can be removed
```suggestion
// Can be expanded with visualisation method if needed.
```

FredrikAugust commented on src/stories/Dashboard.stories.tsx:
```diff
@@ -0,0 +1,11 @@
+import { Story } from '@storybook/react';
+import Dashboard from '../components/organisms/Dashboard';
+
+export const Primary: React.FC = () => <Dashboard />;
+
+export default {
+    title: 'Component/Dashboard',
+    component: Dashboard,
+} as Meta;
```
Component isn't Dashboard, it's Primary.
@blauks might be able to give you some pointers here.

blauks replied:
Hey, @blauks here! According to the docs you are supposed to use the imported component, and not a story. So this can be resolved @emilom

FredrikAugust replied:
Noted! Good catch. 

Close this when you see it @emilom.

FredrikAugust replied:
I suppose it's there to tell you which component the story is describing?

FredrikAugust commented on src/queries/metApi.ts:
```diff
@@ -0,0 +1,24 @@
+import { gql } from '@apollo/client';
+
+//Format: What where
```
What does this mean?

emilom replied:
Describing the naming format for queries. What = WEATHER and where = MET_API.

FredrikAugust replied:
OK. Could it be expressed in a clearer way?

E.g.

```tsx
// Naming convention:
// <RESOURCE>_<SOURCE>
```

FredrikAugust replied:
By the way, the intention behind these comments is not to tell you what to do, so don't feel obligated to incorporate them. It's rather to express my thoughts and reactions.

If you feel the suggestions are in any way suboptimal, don't hesitate to provide counter-arguments or initiate a debate (med vennlig hilsen Fredrik Solvang).

FredrikAugust commented on src/index.tsx:
```diff
@@ -1,6 +1,23 @@
 
-render(<Filler />, document.getElementById('root'));
+const client = new ApolloClient({
+    uri: '/graphql',
```
This should be read from an environment variable

FredrikAugust replied:
And thus also document the environment variable in README+wiki.

FredrikAugust commented on src/index.tsx:
```diff
@@ -1,6 +1,23 @@
+            <Dashboard />
+        </ApolloProvider>
+    );
+};
```
Move this to its own file.

FredrikAugust commented on src/index.tsx:
```diff
@@ -1,6 +1,23 @@
 import { render } from 'react-dom';
+import { ApolloClient } from '@apollo/client';
+import { ApolloProvider } from '@apollo/client';
+import { cache } from './cache/cache';
```
Do you need a folder for cache with a file called cache?
If it's the "root" cache it's common to name it `index.ts`

emilom committed:
Update src/types/dashboard.ts

Co-authored-by: Jonathan Brooks <brooks.jonathan1@gmail.com>

FredrikAugust commented on src/components/organisms/Dashboard.tsx:
```diff
@@ -0,0 +1,70 @@
+            <Pane
+                width="80%"
+                display="grid"
+                gridTemplateColumns="1fr 1fr 1fr 1fr 1fr 1fr" //Smaller screen size should have 1f 1fr 1f
```
You can also use `repeat(n, 1fr)` where `n` is the number of columns.

FredrikAugust replied:
At least you can do that in CSS, so I believe it should work here as well.

emilom replied:
I tried, but it didn’t work with evergreen. It might be possible but just passing the string  "repeat(6, 1fr)" did not work.

FredrikAugust commented on src/components/organisms/Dashboard.tsx:
```diff
@@ -0,0 +1,70 @@
+
+const Dashboard: React.FC = () => {
+    //Apollo local state
+    const dashboardItems = useReactiveVar(dashboardItemsVar);
```
Due to this, you should wrap your storybook template in a provider as well.

@blauks knows more about this (regarding templates and exports).

blauks replied:
Done

FredrikAugust commented on src/components/molecules/DataWrapper.tsx:
```diff
@@ -0,0 +1,22 @@
+};
+
+const DataWrapper: React.FC<DataWrapperProps> = ({ children, query }) => {
+    const [formattedData, setFormattedData] = useState<TimeEntry[] | undefined>(undefined);
```
```suggestion
    const [formattedData, setFormattedData] = useState<TimeEntry[]>([]);
```

FredrikAugust replied:
Default to empty formattedData instead might be easier to type. Just a suggestion.

If not, I believe you can omit `| undefinned` and the `undefined` in the argument list, and it will default to `null`, which I believe is more semantically correct here.

FredrikAugust commented on src/components/molecules/DataWrapper.tsx:
```diff
@@ -0,0 +1,22 @@
+
+const DataWrapper: React.FC<DataWrapperProps> = ({ children, query }) => {
+    const [formattedData, setFormattedData] = useState<TimeEntry[] | undefined>(undefined);
+    const { loading, error, data } = useQuery(query);
+
+    useEffect(() => {
+        if (data) setFormattedData(formatMetData(data));
+    }, [data]);
+
+    return <>{children({ data: formattedData, loading, error })}</>;
```
What about handling `loading` and `error` in this level instead of passing it down?

That way you would remove one area of responsibility from the child.

FredrikAugust commented on src/components/molecules/DataWrapper.tsx:
```diff
@@ -0,0 +1,22 @@
+    const [formattedData, setFormattedData] = useState<TimeEntry[] | undefined>(undefined);
+    const { loading, error, data } = useQuery(query);
+
+    useEffect(() => {
+        if (data) setFormattedData(formatMetData(data));
+    }, [data]);
+
+    return <>{children({ data: formattedData, loading, error })}</>;
```
Since `data` will only change once, you could structure it like this (presuming you handle `loading`/`error` in this component as well)

```tsx
if (loading) {
  return <Loading />;
}

if (error) {
  return <Error error={error} />;
}

return children(data);
```

This would make typing the child a lot easier.

emilom replied:
How do you return just `children(data)`. I can’t get it to work without wrapping it in `<>{children(data)}</>`

FredrikAugust replied:
I don't know, I just made something up.

You can try
- looking at the stack trace. what went wrong?
- what type did it produce, and what does it expect?
- googling the issue
- taking a look at the typing in other examples

FredrikAugust replied:
For reference, you're wrapping the child function invocation in a fragment. To understand why it doesn't work without, it might be helpful to understand what this does. 

It might also be helpful to research these kinds of issues before asking in the code review. If you're still stuck after some time, you can in #help-programming in Slack. I'd appreciate it if we could (for the most part) try to keep code reviews to discussions rather than programming questions.

FredrikAugust commented on src/components/molecules/DataWrapper.tsx:
```diff
@@ -0,0 +1,22 @@
+    const { loading, error, data } = useQuery(query);
+
+    useEffect(() => {
+        if (data) setFormattedData(formatMetData(data));
```
Here you limit this component to only working with `metData`. Could you pass this mapping function in from props?

emilom replied:
That could work. Were thinking about using a switch based on the query to know which format that should be used. Not sure what the best approach is.

FredrikAugust replied:
Generics! They are a perfect fit for this (at least I think so):)

FredrikAugust commented on src/components/atoms/AddComponentButton.tsx:
```diff
@@ -0,0 +1,30 @@
+import { WEATHER_MET_API } from '../../queries/metApi';
+
+const AddComponentButton: React.FC = () => {
+    const addDashboardItem = () => {
+        dashboardItemsVar([...dashboardItemsVar(), { size: 3, withText: false, query: WEATHER_MET_API }]);
+    };
```
This causes the storybook to crash when you try to click the button. You should either mock it in some way or provide a provider in the storybook example.

FredrikAugust commented on src/index.css:
```diff
@@ -5,4 +5,4 @@ html {
     margin: 0;
-}
\ No newline at end of file
+}
```
funny man with the magic words 🧙🏻 you just outperformed me

FredrikAugust commented on src/components/molecules/DataWrapper.tsx:
```diff
@@ -0,0 +1,22 @@
+
+export type DataWrapperProps = {
+    query: DocumentNode;
+    children: (args: { data: TimeEntry[] | undefined; loading: boolean; error: ApolloError | undefined }) => ReactNode;
```
This limits the DataWrapper to only work with the `ThresholdChart` (only accepting `data: TimeEntry[]`).

I believe this would be a fitting time to introduce ✨ _generics_ ✨ 

In essence, you wish to create to create a `DataWrapper` which takes
- a `query` which resolves to a type `T` (for use in `useQuery<T>(query)`)
- a mapping function mapping from type `T -> U`, where `U` is the input to the child function
_What does this mean?_ Albeit a slightly functional (and specifically Haskell) explanation, it essentially means you have to provide a function (in the props) which takes the **query data type** (`T`, e.g. `MetApiRawData`) and maps it to **some arbitrary output type** (`U`). In your example, you have provided this, only hard-coded. 

`formatMetData` takes the raw data type `T` as input, and produces some output type `U` which is further used by the children (in this example, `TimeEntry[]`). 

Thus we can say that `formatMetData` is a mapping from `T -> U`.

If you further eliminate most data from the child function, so you're left with `children: (data: U) => ReactNode`, this can be typed quite neatly, and you have provided the following;

- a **generic** (:sparkles: :mage:) component which retrieves an arbitrary GraphQL data source and provides it to a child visualisation. Given that you handle the error/loading states in this component, the child component (read: not the child function) can have a prop type like `props: { data: TimeEntry[] }` (note that it's not nullable) as you guarantee that it will be provided with actual data.

The usage would thus look like;

```tsx
<DataWrapper<MetAPIRawData, TimeEntry[]>
  mapping={(entry) => ({ x: entry.x, y: entry.y })}
  query={I_LOVE_THE_MONKEY_HEAD_QUERY}
>
  {(data) => <ThresholdChart data={data} />}
</DataWrapper>
```

FredrikAugust replied:
Further information:

If you don't feel comfortable with generics:
- https://www.typescriptlang.org/docs/handbook/generics.html

- Google "react generic components"
- https://wanago.io/2020/03/09/functional-react-components-with-generic-props-in-typescript/

blauks replied:
Thanks for the good tips! I will definitely look more into generic types:)

FredrikAugust commented on src/components/organisms/Dashboard.tsx:
```diff
@@ -0,0 +1,70 @@
+                                        <DataWrapper query={item.query}>
+                                            {({ data, loading, error }) => (
+                                                <>
+                                                    {loading && <Text>Loading...</Text>}
+                                                    {error && <Text>{error.message}</Text>}
+                                                    {data && (
+                                                        <ThresholdChart
+                                                            data={data}
+                                                            height={parent.height}
+                                                            width={parent.width}
+                                                        />
+                                                    )}
+                                                </>
```
This error handling doesn't belong in the `organism`. It only concerns the data, which is handled by `molecule`.

FredrikAugust commented on src/components/organisms/Dashboard.tsx:
```diff
@@ -0,0 +1,70 @@
+                                    {(parent) => (
+                                        <DataWrapper query={item.query}>
+                                            {({ data, loading, error }) => (
+                                                <>
```
Loading, error and data are exclusive states -- i.e. they cannot co-exist, which I believe can be conveyed in a cleaner format.

FredrikAugust replied:
Read molecule comments

FredrikAugust commented on src/components/organisms/Dashboard.tsx:
```diff
@@ -0,0 +1,70 @@
+    return (
+        <Pane width="100%" height="100%" display="flex" flexDirection="column" alignItems="center" paddingBottom="20px">
+            <Pane width="80%" height="100px" position="relative">
+                <Select position="absolute" bottom="20px" width="200px">
```
Is it possible to avoid using pixels?

`rem`s are generally better if you don't need very specific control over an elements size.

blauks replied:
Probably!

FredrikAugust commented on src/components/atoms/AddComponentButton.tsx:
```diff
@@ -0,0 +1,30 @@
+            cursor="pointer"
+            width="100%"
+            height="100%"
+            gridColumn={'span 2'}
```
```suggestion
            gridColumn="span 2"
```

FredrikAugust commented on src/components/atoms/AddComponentButton.tsx:
```diff
@@ -0,0 +1,30 @@
+
+const AddComponentButton: React.FC = () => {
+    const addDashboardItem = () => {
+        dashboardItemsVar([...dashboardItemsVar(), { size: 3, withText: false, query: WEATHER_MET_API }]);
```
Ref. discussion with Jonathan, do you need `withText`? Or can you infer it from the content of the data/components?

emilom replied:
changed the type to `description?: string`

FredrikAugust commented on src/cache/cache.ts:
```diff
@@ -0,0 +1,18 @@
+import { DashboardItemInfo } from '../types/dashboard';
+
+export const cache = new InMemoryCache({
+    typePolicies: {
+        Query: {
+            fields: {
+                dashboardItems: {
+                    read() {
+                        return dashboardItemsVar();
+                    },
+                },
+            },
+        },
+    },
+});
```
Cool! Never seen it in use before.

FredrikAugust commented:
UI Tests denied as it's broken. Ref other comments.

FredrikAugust commented on src/queries/metApi.ts:
```diff
@@ -0,0 +1,24 @@
+            }
+        }
+    }
+`;
```
I'd also recommend typing the result in this file, as @blauks witnessed in ekskom.

So you export the GraphQL query as well as 

```ts
export type WeatherMetApi = {
  // type type type
}
```

FredrikAugust commented:
It also appears that this PR broke the docker registry configuration.
It would be great if you could take a look at this.
If you cannot find any reason for why this is happening, you can replace

```diff
- # RUN yarn build
- RUN npx react-scripts build
+ RUN yarn build
```

emilom committed:
Update src/components/atoms/AddComponentButton.tsx

Co-authored-by: Fredrik August Madsen-Malmo <fredrik.malmo@icloud.com>

Commit:
Fix dashboard type

Commit:
Change query naming convention description

FredrikAugust commented:
By the way, it would be nice if you could write some more specific commit messages. This makes it a bit easier to keep track of what's happening during the review.

E.g. `Fix dashboard type` could be `Use optional paragraph instead of withText in DashboardItemInfo`

It's worth spending a short minute coming up with a good commit message.

Commit:
Change DataWrapper to handle loading/error instead of passing it down

Commit:
Change cache.ts to index.ts

Commit:
Remove dasboardList.ts as it was not in use

FredrikAugust commented on src/components/organisms/Dashboard.tsx:
```diff
@@ -47,19 +47,15 @@ const Dashboard: React.FC = () => {
-                                                </>
-                                            )}
+                                            {({ data }) =>
+                                                data.length !== 0 && (
```
Is this a necessary check?

In the case where it's zero you are we are now rendering _nothing_ instead of an empty chart.

FredrikAugust replied:
You could also perform a check in the `DataWrapper` component to see if data is a falsy value.

emilom replied:
ThresholdChart does not work with an empty array, so it will throw an error

FredrikAugust replied:
OK. Perhaps it would be wise to show an error if the data array is empty?

If you don't have capacity to fix this in this PR, you can create an issue for it.

emilom replied:
Sounds good. I'll create an issue as the design for error messages and loading is due today.

emilom replied:
new issues #50

FredrikAugust commented on src/components/molecules/DataWrapper.tsx:
```diff
@@ -1,22 +1,29 @@
     }, [data]);
 
-    return <>{children({ data: formattedData, loading, error })}</>;
+    if (loading) {
+        return <p>Loading...</p>; //Change to loading atom
+    }
+    if (error) {
+        return <p>Error :/</p>; //Change to error atom
+    }
```
Create issues for this if it's not going to be addressed in this PR.

FredrikAugust replied:
Ties up to #32 which is scheduled for completion today.

emilom replied:
new issues #50

FredrikAugust commented on src/components/molecules/DataWrapper.tsx:
```diff
@@ -1,22 +1,29 @@
+        return <p>Loading...</p>; //Change to loading atom
+    }
+    if (error) {
+        return <p>Error :/</p>; //Change to error atom
```
Might be helpful to render `error.message` for the time being while a proper atom is being constructed.

FredrikAugust reviewed (dismissed):
I'm unable to comment on the line, but it's not recommended to use indexes as keys.

> We don’t recommend using indexes for keys if the order of items may change. This can negatively impact performance and may cause issues with component state. Check out Robin Pokorny’s article for an in-depth explanation on the negative impacts of using an index as a key. If you choose not to assign an explicit key to list items then React will default to using indexes as keys.

https://reactjs.org/docs/lists-and-keys.html#keys

FredrikAugust commented on src/components/molecules/DataWrapper.tsx:
```diff
@@ -0,0 +1,22 @@
+    const { loading, error, data } = useQuery(query);
+
+    useEffect(() => {
+        if (data) setFormattedData(formatMetData(data));
```
`formatMetData` (or the equivalent function (mapping `T->U`)) can also be used in the `useQuery` call;

https://www.apollographql.com/docs/react/data/queries/#options

Look at `onCompleted`

FredrikAugust replied:
Note that this isn't relevant if you remove the `setFormattedData` call, but it can be useful to note.

FredrikAugust committed:
Make DataWrapper generic

DataWrapper now takes two generic parameters — input data type and output data type. Also takes a mapping function as a prop.

blauks committed:
Move App from index to seperate file

FredrikAugust commented on src/components/organisms/Dashboard.tsx:
```diff
@@ -0,0 +1,79 @@
+
+    return (
+        <Pane width="100%" height="100%" display="flex" flexDirection="column" alignItems="center" paddingBottom="20px">
+            <Pane width="80%" height="100px" position="relative">
+                <Select position="absolute" bottom="20px" width="200px">
```
Why do these two use relative/absolute positioning? 

Is it possible to position them using grid/flex instead?

emilom replied:
should be possible 👍

emilom replied:
just wanted to see what it would look like

FredrikAugust commented:
Alright.

If it’s just poking around to see how things react, try to avoid pushing it for review.

> On 2 Mar 2021, at 19:35, emilom <notifications@github.com> wrote:
> 
> 
> @emilom commented on this pull request.
> 
> In src/components/organisms/Dashboard.tsx <https://github.com/IT2901-SINTEF01/frontend/pull/49#discussion_r585812058>:
> 
> > +            <Pane width="80%" height="100px" position="relative">
> +                <Select position="absolute" bottom="20px" width="200px">
> just wanted to see what it would look like
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/IT2901-SINTEF01/frontend/pull/49#discussion_r585812058>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB2V2G45WKVU2TDM4X2DHPLTBUVX7ANCNFSM4YO2INFQ>.
>

blauks committed:
Add enviroment file for graphql endpoint

Commit:
Change styling in Dashboard to flex and rem

Commit:
Merge branch '30-create-dashboard' of https://github.com/IT2901-SINTEF01/frontend into 30-create-dashboard

Commit:
Merge branch '30-create-dashboard' of https://github.com/IT2901-SINTEF01/frontend into 30-create-dashboard

Commit:
Merge branch '30-create-dashboard' of https://github.com/IT2901-SINTEF01/frontend into 30-create-dashboard

blauks committed:
Mock dashboard data in storybook

Commit:
Move return type backend from types to query folder

blauks committed:
Update Dockerfile

blauks committed:
Merge branch '30-create-dashboard' of github.com:IT2901-SINTEF01/frontend into 30-create-dashboard

blauks commented:
> It also appears that this PR broke the docker registry configuration.
> It would be great if you could take a look at this.
> If you cannot find any reason for why this is happening, you can replace
> 
> ```diff
> - # RUN yarn build
> - RUN npx react-scripts build
> + RUN yarn build
> ```

This did not work. I will google some more

blauks committed:
Try to ignore eslintrc and prettierrc in docker

FredrikAugust commented on .dockerignore:
```diff
@@ -1,6 +1,8 @@
 node_modules
 .eslint*
+.eslintrc.js
 .prettier*
+.prettierrc.js
```
The two `.js` files are already handled by the wildcard-ignores.

blauks committed:
Remove explicit typing of generics in dashboard, as it is not needed

blauks commented:
Small error with threshold chart hehe, no need for UI review right now.

blauks committed:
Fix errors in chart story

blauks committed:
Merge branch 'master' into 30-create-dashboard
